pkg/networkservice/common/connect/server.go

```diff
@@ -65,80 +97,204 @@ func WithDialOptionFactory(factory func(ctx context.Context, request *networkser
+		connections:         map[string]*clientEntry{},
+		dialOptions:         clientDialOptions,
+		dialOptionFactories: dialOptionFactories,
+		dialer:              dialer,
 	}
 }
 
-func (c *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	client, found := c.connectionIDToClientMap.Load(request.GetConnection().GetId())
-	if found {
-		return client.Request(ctx, request)
-	}
-	u := clienturl.ClientURL(ctx)
-	client, found = c.uRLToClientMap.Load(u.String())
-	if found {
-		c.connectionIDToClientMap.LoadOrStore(request.GetConnection().GetId(), client)
-		return client.Request(ctx, request)
-	}
-	// TODO - do something with cancelFunc to clean up uRLToClientMap when we are done with them (maybe ref counting on Close?)
-	clientCtx, _ := context.WithCancel(context.Background())
-	// TODO - fix to accept dialOptions from github.com/networkservicemesh/sdk/tools/security - the options should flow in from the top
-	// TODO - fix to accept a mockable interface we can pass in from the top other than DialContext
-	// TODO - fix to be cautious about schemes
-	dialOptions := c.dialOptions
-	for _, dof := range c.dialOptionFactories {
-		dialOptions = append(dialOptions, dof(ctx, request, u)...)
+func grpcDialer(ctx context.Context, target string, opts ...grpc.DialOption) (conn grpc.ClientConnInterface, cancelFunc func(), err error) {
+	var grpcCon *grpc.ClientConn
+	grpcCon, err = grpc.DialContext(ctx, target, opts...)
+	cancelFunc = func() {
+		if grpcCon != nil {
+			_ = grpcCon.Close()
+		}
 	}
-	cc, err := grpc.DialContext(clientCtx, grpcutils.URLToTarget(u), dialOptions...)
+	return grpcCon, cancelFunc, err
+}
+
+func (c *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	// Find or Create clientEntry, with add pending.
+	ce, deferFunc, err := c.findOrCreateClient(ctx)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Unable to dial %s", u.String())
+		// In case of error, no pending state is propogated.
+		return nil, err
 	}
-	client = c.clientFactory(clientCtx, cc)
-	client, _ = c.uRLToClientMap.LoadOrStore(u.String(), client)
-	client, _ = c.connectionIDToClientMap.LoadOrStore(request.GetConnection().GetId(), client)
-	if _, err = client.Request(ctx, request); err != nil {
+	// Will decrease pending and close connection if required
+	defer deferFunc()
+
+	// We have connection ready,
+	var conn *networkservice.Connection
+	conn, err = ce.client.Request(ctx, request)
+	if err != nil {
+		// We not succeed, just remove connection by id, if it was existing one.
+		<-c.executor.AsyncExec(func() {
+			if request.Connection.Id != "" {
+				// Delete from CE
+				delete(ce.connections, request.Connection.Id)
+				// Delete from global list
+				delete(c.connections, request.Connection.Id)
+			}
+		})
 		return nil, err
 	}
+	// we succeed, update connection and remove pending state
```

Comment:
I don't quite follow why we have pending state...

Reply:
pending state is when we in request, but not have connection yet.
So I'e +1 pending on Any Request/Close, and -1 on defer from calls. And in case I've had connection I've add it to map of connection, and client is not closed/removed until we had connections or pending. I could rename it to activeRequests

Reply:
Removed pending state, use connections instead.